RES-21: Process data for network API

akhilaananthram opened the pull request:
NetworkDataGenerator creates a two files, the first one contains the data in the format that the network API FileRecordStream needs and the other is a mapping of category labels to ids.  It has the option to randomize and/or tokenize as well.

@BoltzmannBrain

BoltzmannBrain commented on fluent/utils/network_data_generator.py:
```diff
@@ -0,0 +1,195 @@
+
+import argparse
+import csv
+import json
```
[this is the preferred method of importing json](https://github.com/numenta/nupic.fluent/blob/master/fluent/models/classification_model.py#L30-33)

BoltzmannBrain commented on fluent/utils/network_data_generator.py:
```diff
@@ -0,0 +1,195 @@
+from fluent.utils.text_preprocess import TextPreprocess
+
+
+class NetworkDataGenerator(object):
```
3 blank lines before class name (or top-level function)

BoltzmannBrain commented on fluent/utils/network_data_generator.py:
```diff
@@ -0,0 +1,195 @@
+    self.fieldNames = ["token", "_sequenceID", "_reset"]
+    self.types = {"token": "str",
+                  "_sequenceID": "int",
+                  "_reset": "bool"}
```
you mean int? This would cause the FileRecordSensor to raise an error...

akhilaananthram replied:
When it writes to the file it writes int, not "int".  That matches the format of the data files in nupic.  I'm a bit unclear on why FileRecordSensor would raise an error

BoltzmannBrain replied:
Check out the [FileRecordSensor docstring](https://github.com/numenta/nupic/blob/master/nupic/data/file_record_stream.py#L134-135). Run the script and look at the csv... the type row of "_reset" specifies bool... this is wrong, and gets caught [here](https://github.com/numenta/nupic/blob/master/nupic/data/file_record_stream.py#L236).

BoltzmannBrain replied:
Similarly, FileRecordStream doesn't accept "str", but rather "string":

    Exception: Invalid file format for "/Users/alavin/nta/nupic.fluent/tempOut.csv" - field type "str" not one of ('string', 'datetime', 'int', 'float', 'bool', 'sdr') 

These are things you should catch in testing the first time around.

BoltzmannBrain commented on fluent/utils/network_data_generator.py:
```diff
@@ -0,0 +1,195 @@
+    """
+    # Update header details
+    for i in xrange(len(categoryIndices)):
+      categoryKey = "_category{}".format(i)
```
:+1: str.format()

BoltzmannBrain commented on fluent/utils/network_data_generator.py:
```diff
@@ -0,0 +1,195 @@
+
+
+  def randomizeData(self):
+    self.preprocesseData = random.shuffle(self.preprocessedData)
```
`random.shuffle` operates in place, not to mention the variable being assigned (a None value) is spelled wrong

akhilaananthram replied:
oh! that explains why it worked then when I called it.  Will fix that.

BoltzmannBrain commented on fluent/utils/network_data_generator.py:
```diff
@@ -0,0 +1,195 @@
+
+    with open(dataOutputFile, 'w') as f:
+      # Header
+      writer = csv.DictWriter(f, fieldnames=self.fieldNames)
```
use [csv_helper.py](https://github.com/numenta/nupic.fluent/blob/master/fluent/utils/csv_helper.py#L88) for generic csv writing/reading; the use of `pandas.read_csv()` is okay though

if you need to make changes to csv_help.py, push them in this PR

akhilaananthram replied:
Sorry. I thought this was supposed to be independent of csv_helper.  I can just use the readCSV method.

BoltzmannBrain replied:
If you're using pandas functionality, for reading into a DataFrame it's better to just stick with the pandas method. For generic csv operations use csv_helper -- that's what the utils functions are there for.

BoltzmannBrain commented on fluent/utils/network_data_generator.py:
```diff
@@ -0,0 +1,195 @@
+
+    with open(categoriesOutputFile, 'w') as f:
+      stringOfMap = json.dumps(self.categoryToId)
+      f.write(stringOfMap)
```
write this to a json; if you're going to dump it, write it to a json, otherwise the format comes out terrible in a csv

akhilaananthram replied:
I have been saving it to a json file.  Do you want me to add an assert to confirm that the user gave the path to a json file?

BoltzmannBrain replied:
not an assert, raise a TypeError

and even better, specify a default value in the out file args... `network_experiment/categories.json` and `network_experiment/data.csv'

BoltzmannBrain commented on fluent/utils/network_data_generator.py:
```diff
@@ -0,0 +1,195 @@
+    self.fieldNames = ["token", "_sequenceID", "_reset"]
+    self.types = {"token": "str",
+                  "_sequenceID": "int",
+                  "_reset": "bool"}
```
int

BoltzmannBrain commented on fluent/utils/network_data_generator.py:
```diff
@@ -0,0 +1,195 @@
+
+
+
+def parse_args():
```
fix the spacing in this function, and add single character flags (`--fooBarBaz` --> `-f`) for the req'd args; it's a pain to write out the full arg name each time

see experiment.py for an example of how to do these things

BoltzmannBrain commented:
@akhilaananthram done w/ this round of review. Please proofread/cleanup before pushing code, and add some unit tests please.

BoltzmannBrain commented:
@akhilaananthram another bug: the add'l category columns "_category1" and "_category2" get auto-populated with 1s when they should be blank, e.g. the first sequence of the sample training data.

akhilaananthram commented:
@BoltzmannBrain: added unit tests

BoltzmannBrain commented on fluent/utils/network_data_generator.py:
```diff
@@ -0,0 +1,218 @@
+  parser.add_argument("--f", type=str, required=True,
+    help="path to input file. REQUIRED")
+  parser.add_argument("--o", default="network_experiment/data.csv", type=str,
+    required=True, help="File to write processed data to. REQUIRED")
```
the user should not be req'd to specify an arg that has a default value; this renders the default useless

BoltzmannBrain commented:
@akhilaananthram 
Running network_data_generator.py fails because there is no check to make sure the out file exists for writing. Again, you need to test your code thoroughly. Code in review should not fail, especially this easily.

And it looks like you didn't consider the argparse example in experiment.py at all like I suggested...

akhilaananthram commented:
@BoltzmannBrain: I added the checks to make sure all files provided exist and updated the argparse.  Last time my changes to argparse included a shorter form like experiment.py.  This time it includes both.  I also added the randomization test you suggested.

BoltzmannBrain commented on tests/unit/utils/network_data_generator_test.py:
```diff
@@ -0,0 +1,233 @@
+
+  def testSaveDataIncorrectType(self):
+    ndg = NetworkDataGenerator()
+    filename = "../../../data/network_data_generator/multi_sample.csv"
```
Looking for a file like this assumes the user is testing from a specific directory, which will often be an incorrect assumption. Use `os.path.dirname(os.path.realpath(__file__))` to start from the file's directory and then navigate to the test data file.

BoltzmannBrain commented:
@akhilaananthram please fix the unit tests per my comments on the file paths. I'll fix the rest (argparse) afterwards.

akhilaananthram commented:
@BoltzmannBrain: fixed the file paths
